From 538b4b39c05b70b6907818f72934ce6546fef229 Mon Sep 17 00:00:00 2001
From: Loic Poulain <loic.poulain@intel.com>
Date: Wed, 3 Feb 2016 02:50:16 +0100
Subject: [PATCH 22/24] cpuset: Add allow_attach hook for cpusets on android

We need to add extend permission check for the 'cpuset' subsystem.
Indeed, system_server (non-root) process need to place some other
process in a cpuset.

allow_attach is called prior to moving a task into a cgroup. if the
subsystem returns an error, this will abort the attach operation.

This patch adds the 'allow_attach' handler for the 'cpuset' cgroup to
allow non-root processes to add processes it can ptrace to a 'cpuset'
cgroup.

Change-Id: I090fffd1dc34c2ce2d7a21f44a325dec6e56c424
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-16183
Signed-off-by: Damien Vagner <damienx.vagner@intel.com>
Signed-off-by: Loic Poulain <loic.poulain@intel.com>
Signed-off-by: mark gross <mark.gross@intel.com>
Reviewed-on: https://android.intel.com:443/462951
---
 kernel/cpuset.c | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/kernel/cpuset.c b/kernel/cpuset.c
index 7b4530b..3f39104 100644
--- a/kernel/cpuset.c
+++ b/kernel/cpuset.c
@@ -42,6 +42,7 @@
 #include <linux/namei.h>
 #include <linux/pagemap.h>
 #include <linux/proc_fs.h>
+#include <linux/ptrace.h>
 #include <linux/rcupdate.h>
 #include <linux/sched.h>
 #include <linux/seq_file.h>
@@ -2018,6 +2019,26 @@ static void cpuset_css_free(struct cgroup_subsys_state *css)
 	kfree(cs);
 }
 
+/*
+ * Check if the current process is allowed to move a task to a cpuset
+ * cgroup by checking if the process is ptrace-able by current.
+ */
+static int cpuset_allow_attach(struct cgroup_subsys_state *css,
+			       struct cgroup_taskset *tset)
+{
+	const struct cred *tcred;
+	struct task_struct *task;
+
+	cgroup_taskset_for_each(task, css, tset) {
+		tcred = __task_cred(task);
+
+		if ((current != task) &&
+			!ptrace_may_access(task, PTRACE_MODE_ATTACH))
+			return -EACCES;
+	}
+	return 0;
+}
+
 struct cgroup_subsys cpuset_subsys = {
 	.name = "cpuset",
 	.css_alloc = cpuset_css_alloc,
@@ -2027,6 +2048,7 @@ struct cgroup_subsys cpuset_subsys = {
 	.can_attach = cpuset_can_attach,
 	.cancel_attach = cpuset_cancel_attach,
 	.attach = cpuset_attach,
+	.allow_attach = cpuset_allow_attach,
 	.subsys_id = cpuset_subsys_id,
 	.base_cftypes = files,
 	.early_init = 1,
-- 
1.9.1

